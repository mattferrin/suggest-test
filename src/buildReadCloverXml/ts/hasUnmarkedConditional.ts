import { MeowOutput } from "../../bin/ts/cli";
import { FileArray } from "./summarizedResult";
import { Untrusted } from "./Untrusted";

export function hasUnmarkedConditional(
  input: MeowOutput,
  file: Untrusted<FileArray[number]>
): boolean {
  if (input.flags.includeConditionals) {
    const hasConditional =
      file?.metrics?.some((metric) => {
        return Number.parseInt(metric?.$?.conditionals ?? "0") > 0;
      }) ?? false;

    return (
      hasConditional &&
      (file?.line?.every((line) => {
        return line?.$?.type !== "cond";
      }) ??
        false)
    );
  } else {
    return false;
  }
}
