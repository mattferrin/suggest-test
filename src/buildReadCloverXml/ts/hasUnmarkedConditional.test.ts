import { hasUnmarkedConditional } from "./hasUnmarkedConditional";

type Input = Parameters<typeof hasUnmarkedConditional>[0];
type ProjectPackage = Parameters<typeof hasUnmarkedConditional>[1];

function snap(input: {
  stabileSortId: string;
  comment: string | null;
  input: Input;
  file: ProjectPackage;
}): void {
  it(input.stabileSortId, () => {
    /** setup mocks */

    /** unit under test */
    const result = hasUnmarkedConditional(input.input, input.file);

    /** assertions */
    expect({
      [`${input.comment ?? "TODO: comment"}`]: {
        result,
      },
    }).toMatchSnapshot();
  });
}

snap({
  stabileSortId: "f823b49a-cfdd-40c9-8dac-5f7a9975d0c6",
  comment:
    "`true` WHEN conditional exists AND NO line data AND conditionals included",
  input: { flags: { includeConditionals: true } } as Input,
  file: {
    metrics: [{ $: { conditionals: "1" } }],
    line: [{ $: { type: "NOT cond" as any } }],
  } as ProjectPackage,
});
snap({
  stabileSortId: "424ba948-963c-46d0-b32b-07ade51c9f31",
  comment:
    "`false` WHEN conditional exists AND NO line data AND conditionals excuded",
  input: { flags: { includeConditionals: false } } as Input,
  file: {
    metrics: [{ $: { conditionals: "1" } }],
    line: [{ $: { type: "NOT cond" as any } }],
  } as ProjectPackage,
});

snap({
  stabileSortId: "e9609fe2-44db-47b5-91d3-27831468a361",
  comment: "`false` WHEN NO conditionals",
  input: { flags: { includeConditionals: true } } as Input,
  file: {
    metrics: [{ $: { conditionals: "0" } }],
    line: [{ $: { type: "NOT cond" as any } }],
  } as ProjectPackage,
});

snap({
  stabileSortId: "b2f223ca-cd12-4ce6-b5cb-dc0c62642a17",
  comment:
    "`false` WHEN conditional line data exists AND conditionals included",
  input: { flags: { includeConditionals: true } } as Input,
  file: {
    metrics: [{ $: { conditionals: "1" } }],
    line: [{ $: { type: "cond" as any } }],
  } as ProjectPackage,
});

snap({
  stabileSortId: "ca766aa2-a8c8-4b12-9700-b1b771ae11c7",
  comment: "`false` WHEN null input",
  input: { flags: { includeConditionals: true } } as Input,
  file: null,
});
snap({
  stabileSortId: "e5a1d9b8-4a9d-4776-98e8-6c11ee37095c",
  comment: "`false` WHEN undefined input",
  input: { flags: { includeConditionals: true } } as Input,
  file: undefined,
});

snap({
  stabileSortId: "7fcd5863-dde8-4427-a2d7-c21195086170",
  comment:
    "`false` WHEN NO line data AND conditionals included AND undefined metrics item",
  input: { flags: { includeConditionals: true } } as Input,
  file: {
    metrics: [undefined],
    line: [{ $: { type: "NOT cond" as any } }],
  } as ProjectPackage,
});
snap({
  stabileSortId: "e18a4ddd-68e4-4bb2-ba82-622bd574d66b",
  comment:
    "`false` WHEN NO line data AND conditionals included AND null metrics item",
  input: { flags: { includeConditionals: true } } as Input,
  file: {
    metrics: [null],
    line: [{ $: { type: "NOT cond" as any } }],
  } as ProjectPackage,
});

snap({
  stabileSortId: "06c24ded-fd1e-44e7-aca5-323622f883ca",
  comment:
    "`false` WHEN NO line data AND conditionals included AND undefined metrics",
  input: { flags: { includeConditionals: true } } as Input,
  file: {
    metrics: undefined,
    line: [{ $: { type: "NOT cond" as any } }],
  } as ProjectPackage,
});
snap({
  stabileSortId: "aff17e98-44c1-456f-b06e-e78f221a13d9",
  comment:
    "`false` WHEN NO line data AND conditionals included AND null metrics",
  input: { flags: { includeConditionals: true } } as Input,
  file: {
    metrics: null,
    line: [{ $: { type: "NOT cond" as any } }],
  } as ProjectPackage,
});

snap({
  stabileSortId: "f823b49a-cfdd-40c9-8dac-5f7a9975d0c6",
  comment: "result `false` WHEN line undefined",
  input: { flags: { includeConditionals: true } } as Input,
  file: {
    metrics: [{ $: { conditionals: "1" } }],
    line: undefined,
  } as ProjectPackage,
});
snap({
  stabileSortId: "f823b49a-cfdd-40c9-8dac-5f7a9975d0c6",
  comment: "result `false` WHEN line null",
  input: { flags: { includeConditionals: true } } as Input,
  file: {
    metrics: [{ $: { conditionals: "1" } }],
    line: null,
  } as ProjectPackage,
});
