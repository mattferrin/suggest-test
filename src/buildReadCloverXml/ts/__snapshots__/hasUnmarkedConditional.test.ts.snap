// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`06c24ded-fd1e-44e7-aca5-323622f883ca 1`] = `
{
  "\`false\` WHEN NO line data AND conditionals included AND undefined metrics": {
    "result": false,
  },
}
`;

exports[`7fcd5863-dde8-4427-a2d7-c21195086170 1`] = `
{
  "\`false\` WHEN NO line data AND conditionals included AND undefined metrics item": {
    "result": false,
  },
}
`;

exports[`424ba948-963c-46d0-b32b-07ade51c9f31 1`] = `
{
  "\`false\` WHEN conditional exists AND NO line data AND conditionals excuded": {
    "result": false,
  },
}
`;

exports[`aff17e98-44c1-456f-b06e-e78f221a13d9 1`] = `
{
  "\`false\` WHEN NO line data AND conditionals included AND null metrics": {
    "result": false,
  },
}
`;

exports[`b2f223ca-cd12-4ce6-b5cb-dc0c62642a17 1`] = `
{
  "\`false\` WHEN conditional line data exists AND conditionals included": {
    "result": false,
  },
}
`;

exports[`ca766aa2-a8c8-4b12-9700-b1b771ae11c7 1`] = `
{
  "\`false\` WHEN null input": {
    "result": false,
  },
}
`;

exports[`e5a1d9b8-4a9d-4776-98e8-6c11ee37095c 1`] = `
{
  "\`false\` WHEN undefined input": {
    "result": false,
  },
}
`;

exports[`e18a4ddd-68e4-4bb2-ba82-622bd574d66b 1`] = `
{
  "\`false\` WHEN NO line data AND conditionals included AND null metrics item": {
    "result": false,
  },
}
`;

exports[`e9609fe2-44db-47b5-91d3-27831468a361 1`] = `
{
  "\`false\` WHEN NO conditionals": {
    "result": false,
  },
}
`;

exports[`f823b49a-cfdd-40c9-8dac-5f7a9975d0c6 1`] = `
{
  "\`true\` WHEN conditional exists AND NO line data AND conditionals included": {
    "result": true,
  },
}
`;

exports[`f823b49a-cfdd-40c9-8dac-5f7a9975d0c6 2`] = `
{
  "result \`false\` WHEN line undefined": {
    "result": false,
  },
}
`;

exports[`f823b49a-cfdd-40c9-8dac-5f7a9975d0c6 3`] = `
{
  "result \`false\` WHEN line null": {
    "result": false,
  },
}
`;
